### ISSUE TITLE
REQUEST FOR COMMENTS: How should DnsControl manage MAC addresses?

### ISSUE THREAD

**Issue by tlimoncelli** [StackExchange/dnscontrol#990]
I would like DNSControl to store MAC addresses (ethernet Media Access Control; not Apple Mac) and output them in a format that could be fed into other systems.

Here's the syntax I'm thinking of:

```
    A("myhost", "10.10.10.10", MAC("1:0:5f:7e:1f:fa") ),
```

MAC would return meta info:

```
          { mac_addr: "1:0:5f:7e:1f:fa" }
```

Ideas about the output:

1. Output all the entries into a macaddress.txt file that is just "macaddr IP hostname" (similar to the format of /etc/ethers)
2. Provide a way to filter what does/doesn't go into that file.  Not sure what form that would take.

Providers:

Maybe rather than outputting a file, we should have providers.  It would be nice (for example) to have incremental updates of an Windows DHCP server.  That said, a PowerShell script could probably do the job just as well.

**Comment by patschi** [comment]
Just my two cents...
I don't know the exact use-case you have for this particular feature, but what about the idea of having a more generic approach like "metadata"? Then this would not just limited to MAC addresses (without making new functions/types for each time), rather it could be used for many different scenarios.

Thinking about something like:
```javascript
A( "myhost", "10.10.10.10", META("mac_addr", "1:0:5f:7e:1f:fa") ),
````

This adds metadata with key `mac_addr` of value `1:0:5f:7e:1f:fa` to the record object. In the backend this could be simply an array (as well as on the JavaScript-side). Using `for()` and custom code in JS this could then be read and used for all crazy things.

Benefits might be:
1. We have one generic type `metadata`. Not requiring of individual/special types in the future.
2. As there's one generic type, this should also be way handier in regard to maintenance.
3. More intelligent logic (like validating MAC addresses, writing them in a specific format, etc) can be done using custom JS functions in the individual setup. This would keep the core dnscontrol code "cleaner" and less to take care in regard to maintenance.

In regards to the `macaddress.txt` file: This would mean that the JS code must be able to write/read files from the filesystem? If so, then that would sound great - also thought about implementing that someday, but had no real requirement so far.

In regard to "way to filter what goes into the file": That would be something which could be as well done in the JS logic, as suggested above.

**Comment by tlimoncelli** [comment]
Meta data:

This already exists, just in a different form.  Any `{ key: value }` after the usual arguments (for an A record: label and target; for an MX record: label, priority, target; etc.) is recorded in `RecordConfig.Metadata` map (`map[string]string`).

This is how, for example, CloudFlare's "orange cloud" flag is stored:

    A("www", "10.1.2.3", { cloudflare_proxy: "on" })

Domains have metadata too, as long as it is after the name and registrar:

    D("example.com", REG_NONE, { cloudflare_proxy_default: 'on' })

As do providers:

```
var DSP_MYBIND = NewDnsProvider("bind", "BIND", {
    default_ns: [ "ns1.serverfault.com.", "ns2.serverfault.com." ],
});
```

Macros like `CF_PROXY_ON` and other features are just JS objects (`{ foo: "bar" }`) which are processed later by providers or other tools.

So, basically the proposal is:

    A("myhost", "10.10.10.10", MAC("1:0:5f:7e:1f:fa") ),
    // same as:
    A("myhost", "10.10.10.10", { mac_addr: "1:0:5f:7e:1f:fa" }),

**Comment by tlimoncelli** [comment]
Closing. There's little to no interest in using dnscontrol to manage DHCP.
